openmdao/core/vec_wrapper.py

```diff
@@ -690,9 +690,55 @@ def _get_flattened_sizes(self):
         """
         return [[(n, m['size']) for n, m in self._get_vecvars()]]
 
+    def distance_along_vector_to_limit(self, alpha, duvec):
+        """ Returns a new alpha so that new_u = current_u + alpha*duvec does
+        not violate any `lower` or `upper` limits if specified.
+
+        Args
+        -----
+        alpha: float
+            Initial value for step in gradient direction.
+        duvec: `Vecwrapper`
+            Direction to apply step. generally the gradient.
+
+        Returns
+        --------
+        float
+            New step size, backtracked to prevent violation."""
+
+        # A single index of the gradient can be zero, so we want to suppress
+        # the warnings from numpy.
+        old_warn = numpy.geterr()
+        numpy.seterr(divide='ignore')
+
+        new_alpha = alpha
+        for name, meta in iteritems(self):
+
+            if 'remote' in meta:
+                continue
+
+            val = self[name]
+
+            upper = meta.get('upper')
+            if upper is not None:
+                alpha_bound = numpy.min((upper - val)/duvec[name])
+                if alpha_bound >= 0.0:
+                    new_alpha = min(new_alpha, alpha_bound)
+
+            lower = meta.get('lower')
+            if upper is not None:
```

Comment:
good catch, i need to repeat the tests with just lower defined and just upper defined